src/java.base/macosx/native/libjava/java_props_macosx.c

```diff
@@ -247,15 +247,34 @@ void setOSNameAndVersion(java_props_t *sprops) {
         [invoke getReturnValue:&osVer];
 
         NSString *nsVerStr;
-        if (osVer.patchVersion == 0) { // Omit trailing ".0"
-            nsVerStr = [NSString stringWithFormat:@"%ld.%ld",
-                    (long)osVer.majorVersion, (long)osVer.minorVersion];
-        } else {
-            nsVerStr = [NSString stringWithFormat:@"%ld.%ld.%ld",
-                    (long)osVer.majorVersion, (long)osVer.minorVersion, (long)osVer.patchVersion];
+        // Copy out the char* if running on version other than 10.16 Mac OS (10.16 == 11.x)
+        if (!((long)osVer.majorVersion == 10 &&
+                   (long)osVer.minorVersion >= 16)) {
+            if (osVer.patchVersion == 0) { // Omit trailing ".0"
+                nsVerStr = [NSString stringWithFormat:@"%ld.%ld",
+                        (long)osVer.majorVersion, (long)osVer.minorVersion];
+            } else {
+                nsVerStr = [NSString stringWithFormat:@"%ld.%ld.%ld",
+                        (long)osVer.majorVersion, (long)osVer.minorVersion, (long)osVer.patchVersion];
+            }
+            // Copy out the char*
+            osVersionCStr = strdup([nsVerStr UTF8String]);
+        } else if (getenv("SYSTEM_VERSION_COMPAT") == NULL) {
```

Comment:
If version is 10.16 and `SYSTEM_VERSION_COMPAT` is set, you will fall through to the pre-10.9 Mac OS code fallback. Just checking to see if that's what you intended.

Reply:
FWIW, it seems to work OK using the legacy fallback path (reports 10.16 if I set `SYSTEM_VERSION_COMPAT=1`).

Reply:
The same version string is available from both APIs, reading from the SystemVersion.plist is a bit slower.
It would be clearer to move the checking of SYSTEM_VERSION_COMPAT to the first test (line:252)
so the version info does not need to be read from the files.